[SPARK-20990][SQL] Read all JSON documents in files when multiline mode is on

mgaido91 opened the pull request:
## What changes were proposed in this pull request?

The PR  improves the JSON parsing so that now all the JSON documents in a file are read also when the `multiLine` option is turned on.

## How was this patch tested?

A UT has been added to verify this patch.

mgaido91 committed:
[SPARK-20990][SQL] Read all JSON documents in files when multiline mode is on

AmplabJenkins commented:
Can one of the admins verify this patch?

mgaido91 commented:
cc @gatorsmile

gatorsmile commented:
Thanks for fixing it! Will review it this weekend.

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #79938 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79938/testReport)** for PR 18731 at commit [`972a28d`](https://github.com/apache/spark/commit/972a28d44fcd3c457efc4b732ade47c548766a30).

SparkQA commented:
**[Test build #79938 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79938/testReport)** for PR 18731 at commit [`972a28d`](https://github.com/apache/spark/commit/972a28d44fcd3c457efc4b732ade47c548766a30).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79938/
Test FAILed.

gatorsmile commented:
Could you fix the bug?

mgaido91 commented:
I am debugging, thanks.

mgaido91 commented:
The reason of the UT failure is that in these two UTs we are passing invalid JSONs (mind the extra closed curly brace):

- https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L232;
- https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L242

Parsing them correctly is a bad behavior of the current version IMHO.
I am correcting the JSONs in the UTs and rerun all the UTs locally.
@gatorsmile  let me know if this is ok for you, please

mgaido91 committed:
Fixing bad formatted JSONs in two UTs

SparkQA commented:
**[Test build #79963 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79963/testReport)** for PR 18731 at commit [`4da8493`](https://github.com/apache/spark/commit/4da8493b85f6333df3a3fdc26f5c82be7ebb4ccd).

SparkQA commented:
**[Test build #79963 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79963/testReport)** for PR 18731 at commit [`4da8493`](https://github.com/apache/spark/commit/4da8493b85f6333df3a3fdc26f5c82be7ebb4ccd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79963/
Test PASSed.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JacksonParser.scala:
```diff
@@ -347,13 +347,18 @@ class JacksonParser(
+        allRows
       }
     } catch {
       case e @ (_: RuntimeException | _: JsonProcessingException) =>
```
This fix only works when there do not exist an error. However, we need to support different modes for handling corrupted records. For example, `PERMISSIVE` and `DROPMALFORMED`.

mgaido91 replied:
Thanks for your review.
You are definitely right about it, but I am not sure how to deal with these cases. What I mean is that if we have one JSON per line, for instance, you know that each line is either a valid JSON or not, because you know that at the end of the line the JSON must have been ended. In this case, instead, we don't have a delimiter. The only way we have to say that a JSON is finished is to parse it properly. Otherwise we don't know where a document finishes and the next one begins, because we have no "delimiter" for the end of the JSON.
Do you have any suggestion about how we should deal with this? I mean, do you have any idea about which should be the right "split policy" to apply to choose when a document ends?

gatorsmile replied:
A good reference is CSV. You can take a look at the multi-line support in CSV data sources.

gatorsmile commented:
BTW, you also need to add a few error handling test cases to ensure the corrupted record handling works.

mgaido91 committed:
[SPARK-20990][SQL] Split JSONs in document and parse them to respect policies for bad formatted ones.

SparkQA commented:
**[Test build #80251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80251/testReport)** for PR 18731 at commit [`0a2e8c4`](https://github.com/apache/spark/commit/0a2e8c48c0df46e26d685414d0147151d3f7e784).

SparkQA commented:
**[Test build #80251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80251/testReport)** for PR 18731 at commit [`0a2e8c4`](https://github.com/apache/spark/commit/0a2e8c48c0df46e26d685414d0147151d3f7e784).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `    implicit class JsonCharacter(char: Char) `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80251/
Test FAILed.

mgaido91 committed:
making class JsonCharacter private

mgaido91 committed:
Fix nested objects parsing

SparkQA commented:
**[Test build #80271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80271/testReport)** for PR 18731 at commit [`6071daf`](https://github.com/apache/spark/commit/6071dafb1269f54c9246ebe1cdccaad721365971).

SparkQA commented:
**[Test build #80271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/80271/testReport)** for PR 18731 at commit [`6071daf`](https://github.com/apache/spark/commit/6071dafb1269f54c9246ebe1cdccaad721365971).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/80271/
Test PASSed.

mgaido91 commented:
@gatorsmile any feedback on this?
I added the support for all the corrupt record handling modes and I added the relative tests.
Is anything else needed?

mgaido91 committed:
Merge branch 'master' into SPARK-20990

SparkQA commented:
**[Test build #81183 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81183/testReport)** for PR 18731 at commit [`fdd21eb`](https://github.com/apache/spark/commit/fdd21eba2004f345d358d62733f15fd1fefab1f6).

SparkQA commented:
**[Test build #81183 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81183/testReport)** for PR 18731 at commit [`fdd21eb`](https://github.com/apache/spark/commit/fdd21eba2004f345d358d62733f15fd1fefab1f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81183/
Test PASSed.

mgaido91 commented:
The CI error actually is a timeout. I don't know the reason of the timeout or how to retrigger it. If I should do anything please just tell me.
If there is anything I can do for this PR just tell me please.

mgaido91 commented:
AppVeyor build failed because of a timeout. I don't know how to retest it.

gatorsmile commented:
Thanks! Will review it this weekend

mgaido91 committed:
Merge branch 'master' into SPARK-20990

SparkQA commented:
**[Test build #81696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81696/testReport)** for PR 18731 at commit [`a53404e`](https://github.com/apache/spark/commit/a53404e1d852978297af884970ed5f3fc30e6350).

SparkQA commented:
**[Test build #81696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/81696/testReport)** for PR 18731 at commit [`a53404e`](https://github.com/apache/spark/commit/a53404e1d852978297af884970ed5f3fc30e6350).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `public class NettyMemoryMetrics implements MetricSet `
  * `  class ByteArrayConstructor extends net.razorvine.pickle.objects.ByteArrayConstructor `
  * `public class JavaFeatureHasherExample `
  * `sealed trait LogisticRegressionTrainingSummary extends LogisticRegressionSummary `
  * `sealed trait BinaryLogisticRegressionSummary extends LogisticRegressionSummary `
  * `class ClusteringEvaluator @Since(\"2.3.0\") (@Since(\"2.3.0\") override val uid: String)`
  * `  case class ClusterStats(featureSum: Vector, squaredNormSum: Double, numOfPoints: Long)`
  * `class OneVsRest(Estimator, OneVsRestParams, HasParallelism, JavaMLReadable, JavaMLWritable):`
  * `class HasParallelism(Params):`
  * `case class InsertIntoDir(`
  * `case class InsertIntoDataSourceDirCommand(`
  * `case class DescribeColumnCommand(`
  * `case class InsertIntoHiveDirCommand(`

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/81696/
Test FAILed.

mgaido91 commented:
I close this, I will prepare a new PR, cleaner. Thanks
